pkgs/development/compilers/llvm/15/llvm/default.nix

```diff
@@ -0,0 +1,267 @@
+, debugVersion ? false
+, enableManpages ? false
+, enableSharedLibraries ? !stdenv.hostPlatform.isStatic
+, enablePFM ? !(stdenv.isDarwin
+  || stdenv.isAarch64 # broken for Ampere eMAG 8180 (c2.large.arm on Packet) #56245
+  || stdenv.isAarch32 # broken for the armv7l builder
+)
+, enablePolly ? false
+} @args:
+
+let
+  inherit (lib) optional optionals optionalString;
+
+  # Used when creating a version-suffixed symlink of libLLVM.dylib
+  shortVersion = with lib;
+    concatStringsSep "." (take 1 (splitString "." release_version));
+
+in stdenv.mkDerivation (rec {
+  pname = "llvm";
+  inherit version;
+
+  src = runCommand "${pname}-src-${version}" {} (''
+    mkdir -p "$out"
+    cp -r ${monorepoSrc}/cmake "$out"
+    cp -r ${monorepoSrc}/${pname} "$out"
+    cp -r ${monorepoSrc}/third-party "$out"
+  '' + lib.optionalString enablePolly ''
+    cp -r ${monorepoSrc}/polly "$out/llvm/tools"
+  '');
+
+  sourceRoot = "${src.name}/${pname}";
+
+  outputs = [ "out" "lib" "dev" "python" ];
+
+  nativeBuildInputs = [ cmake python3 ]
+    ++ optionals enableManpages [ python3.pkgs.sphinx python3.pkgs.recommonmark ];
+
+  buildInputs = [ libxml2 libffi ]
+    ++ optional enablePFM libpfm; # exegesis
+
+  propagatedBuildInputs = [ ncurses zlib ];
+
+  checkInputs = [ which ];
+
+  patches = [
+    ./gnu-install-dirs.patch
+  ] ++ lib.optional enablePolly ./gnu-install-dirs-polly.patch;
+
+  postPatch = optionalString stdenv.isDarwin ''
+    substituteInPlace cmake/modules/AddLLVM.cmake \
+      --replace 'set(_install_name_dir INSTALL_NAME_DIR "@rpath")' "set(_install_name_dir)" \
+      --replace 'set(_install_rpath "@loader_path/../''${CMAKE_INSTALL_LIBDIR}''${LLVM_LIBDIR_SUFFIX}" ''${extra_libdir})' ""
+  '' + ''
+    # FileSystem permissions tests fail with various special bits
+    substituteInPlace unittests/Support/CMakeLists.txt \
+      --replace "Path.cpp" ""
+    rm unittests/Support/Path.cpp
+    substituteInPlace unittests/IR/CMakeLists.txt \
+      --replace "PassBuilderCallbacksTest.cpp" ""
+    rm unittests/IR/PassBuilderCallbacksTest.cpp
+    rm test/tools/llvm-objcopy/ELF/mirror-permissions-unix.test
+  '' + optionalString stdenv.hostPlatform.isMusl ''
+    patch -p1 -i ${../../TLI-musl.patch}
+    substituteInPlace unittests/Support/CMakeLists.txt \
+      --replace "add_subdirectory(DynamicLibrary)" ""
+    rm unittests/Support/DynamicLibrary/DynamicLibraryTest.cpp
+    # valgrind unhappy with musl or glibc, but fails w/musl only
+    rm test/CodeGen/AArch64/wineh4.mir
+  '' + optionalString stdenv.hostPlatform.isAarch32 ''
+    # skip failing X86 test cases on 32-bit ARM
+    rm test/DebugInfo/X86/convert-debugloc.ll
+    rm test/DebugInfo/X86/convert-inlined.ll
+    rm test/DebugInfo/X86/convert-linked.ll
+    rm test/tools/dsymutil/X86/op-convert.test
+    rm test/tools/gold/X86/split-dwarf.ll
+    rm test/tools/llvm-dwarfdump/X86/prettyprint_types.s
+    rm test/tools/llvm-dwarfdump/X86/simplified-template-names.s
```

Comment:
```suggestion
    rm test/tools/llvm-dwarfdump/X86/simplified-template-names.s
    rm test/CodeGen/RISCV/rv32zbp.ll
    rm test/CodeGen/RISCV/rv64zbp.ll
```

These tests are failing on armv7l-linux and were [removed](https://github.com/llvm/llvm-project/commit/182aa0cbe0cd86dbf91e9d0e647eec736461f0b7#diff-111093152696dbf9fa35b4948b543cdd57b42b149aead45d7864f80f9c0cafa3) on the `main` branch of LLVM.

<details>
    <summary>Log</summary>

    [100%] Running all regression tests
    /nix/store/piihcvqix0r5zwbi1vfd8m34brcnwr44-python3-3.10.9/bin/python3.10 /build/llvm-src-15.0.6/llvm/build/./bin/llvm-lit -svj4 --no-progress-bar /build/llvm-src-15.0.6/llvm/build/utils/lit /build/llvm-src-15.0.6/llvm/build/test
    llvm-lit: /build/llvm-src-15.0.6/llvm/build/utils/lit/tests/lit.cfg:92: warning: Setting a timeout per test not supported. Requires the Python psutil module but it could not be found. Try installing it via pip or via your operating system's package manager.
     Some tests will be skipped and the --timeout command line argument will not work.
    -- Testing: 42794 tests, 4 workers --
    FAIL: LLVM :: CodeGen/RISCV/rv32zbp.ll (14388 of 42794)
    ******************** TEST 'LLVM :: CodeGen/RISCV/rv32zbp.ll' FAILED ********************
    Script:
    --
    : 'RUN: at line 2';   /build/llvm-src-15.0.6/llvm/build/bin/llc -mtriple=riscv32 -verify-machineinstrs < /build/llvm-src-15.0.6/llvm/test/CodeGen/RISCV/rv32zbp.ll    | /build/llvm-src-15.0.6/llvm/build/bin/FileCheck /build/llvm-src-15.0.6/llvm/test/CodeGen/RISCV/rv32zbp.ll -check-prefixes=CHECK,RV32I
    : 'RUN: at line 4';   /build/llvm-src-15.0.6/llvm/build/bin/llc -mtriple=riscv32 -mattr=+experimental-zbp -verify-machineinstrs < /build/llvm-src-15.0.6/llvm/test/CodeGen/RISCV/rv32zbp.ll    | /build/llvm-src-15.0.6/llvm/build/bin/FileCheck /build/llvm-src-15.0.6/llvm/test/CodeGen/RISCV/rv32zbp.ll -check-prefixes=CHECK
    ,RV32ZBP
    --
    Exit Code: 1

    Command Output (stderr):
    --
    /build/llvm-src-15.0.6/llvm/test/CodeGen/RISCV/rv32zbp.ll:2732:17: error: RV32ZBP-NEXT: expected string not found in input
    ; RV32ZBP-NEXT: rev8.h a0, a0
                    ^
    <stdin>:701:9: note: scanning from here
    # %bb.0:
            ^
    <stdin>:702:2: note: possible intended match here
     rev8 a0, a0
     ^
    /build/llvm-src-15.0.6/llvm/test/CodeGen/RISCV/rv32zbp.ll:2758:17: error: RV32ZBP-NEXT: expected string not found in input
    ; RV32ZBP-NEXT: rev8.h a0, a0
                    ^
    <stdin>:714:9: note: scanning from here
    # %bb.0:
            ^
    <stdin>:715:2: note: possible intended match here
     rev8 a0, a0
     ^

    Input file: <stdin>
    Check file: /build/llvm-src-15.0.6/llvm/test/CodeGen/RISCV/rv32zbp.ll

    -dump-input=help explains the following input dump.

    Input was:
    <<<<<<
                 .
                 .
                 .
               696:  .globl bswap_rotr_i32 # -- Begin function bswap_rotr_i32
               697:  .p2align 2
               698:  .type bswap_rotr_i32,@function
               699: bswap_rotr_i32: # @bswap_rotr_i32
               700:  .cfi_startproc
               701: # %bb.0:
    next:2732'0             X error: no match found
               702:  rev8 a0, a0
    next:2732'0     ~~~~~~~~~~~~~
    next:2732'1      ?            possible intended match
               703:  rori a0, a0, 16
    next:2732'0     ~~~~~~~~~~~~~~~~~
               704:  ret
    next:2732'0     ~~~~~
               705: .Lfunc_end63:
    next:2732'0     ~~~~~~~~~~~~~~
               706:  .size bswap_rotr_i32, .Lfunc_end63-bswap_rotr_i32
    next:2732'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
               707:  .cfi_endproc
    next:2732'0     ~~~~~~~~~~~~~~
               708:  # -- End function
    next:2732'0     ~~~~~~~~~~~~~~~~~~~
               709:  .globl bswap_rotl_i32 # -- Begin function bswap_rotl_i32
    next:2732'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
               710:  .p2align 2
    next:2732'0     ~~~~~~~~~~~~
               711:  .type bswap_rotl_i32,@function
    next:2732'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
               712: bswap_rotl_i32: # @bswap_rotl_i32
    next:2732'0     ~~~~~~~~~~~~~~~
               713:  .cfi_startproc
               714: # %bb.0:
    next:2758'0             X error: no match found
               715:  rev8 a0, a0
    next:2758'0     ~~~~~~~~~~~~~
    next:2758'1      ?            possible intended match
               716:  rori a0, a0, 16
    next:2758'0     ~~~~~~~~~~~~~~~~~
               717:  ret
    next:2758'0     ~~~~~
               718: .Lfunc_end64:
    next:2758'0     ~~~~~~~~~~~~~~
               719:  .size bswap_rotl_i32, .Lfunc_end64-bswap_rotl_i32
    next:2758'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
               720:  .cfi_endproc
    next:2758'0     ~~~~~~~~~~~~~~
                 .
                 .
                 .
    >>>>>>

    --

    ********************
    FAIL: LLVM :: CodeGen/RISCV/rv64zbp.ll (14424 of 42794)
    ******************** TEST 'LLVM :: CodeGen/RISCV/rv64zbp.ll' FAILED ********************
    Script:
    --
    : 'RUN: at line 2';   /build/llvm-src-15.0.6/llvm/build/bin/llc -mtriple=riscv64 -verify-machineinstrs < /build/llvm-src-15.0.6/llvm/test/CodeGen/RISCV/rv64zbp.ll    | /build/llvm-src-15.0.6/llvm/build/bin/FileCheck /build/llvm-src-15.0.6/llvm/test/CodeGen/RISCV/rv64zbp.ll -check-prefix=RV64I
    : 'RUN: at line 4';   /build/llvm-src-15.0.6/llvm/build/bin/llc -mtriple=riscv64 -mattr=+experimental-zbp -verify-machineinstrs < /build/llvm-src-15.0.6/llvm/test/CodeGen/RISCV/rv64zbp.ll    | /build/llvm-src-15.0.6/llvm/build/bin/FileCheck /build/llvm-src-15.0.6/llvm/test/CodeGen/RISCV/rv64zbp.ll -check-prefix=RV64ZBP
    --
    Exit Code: 1

    Command Output (stderr):
    --
    /build/llvm-src-15.0.6/llvm/test/CodeGen/RISCV/rv64zbp.ll:2862:17: error: RV64ZBP-NEXT: expected string not found in input
    ; RV64ZBP-NEXT: rev8.w a0, a0
                    ^
    <stdin>:808:9: note: scanning from here
    # %bb.0:
            ^
    <stdin>:809:2: note: possible intended match here
     rev8 a0, a0
     ^
    /build/llvm-src-15.0.6/llvm/test/CodeGen/RISCV/rv64zbp.ll:2905:17: error: RV64ZBP-NEXT: expected string not found in input
    ; RV64ZBP-NEXT: rev8.w a0, a0
                    ^
    <stdin>:821:9: note: scanning from here
    # %bb.0:
            ^
    <stdin>:822:2: note: possible intended match here
     rev8 a0, a0
     ^

    Input file: <stdin>
    Check file: /build/llvm-src-15.0.6/llvm/test/CodeGen/RISCV/rv64zbp.ll

    -dump-input=help explains the following input dump.

    Input was:
    <<<<<<
                 .
                 .
                 .
               803:  .globl bswap_rotr_i64 # -- Begin function bswap_rotr_i64
               804:  .p2align 2
               805:  .type bswap_rotr_i64,@function
               806: bswap_rotr_i64: # @bswap_rotr_i64
               807:  .cfi_startproc
               808: # %bb.0:
    next:2862'0             X error: no match found
               809:  rev8 a0, a0
    next:2862'0     ~~~~~~~~~~~~~
    next:2862'1      ?            possible intended match
               810:  rori a0, a0, 32
    next:2862'0     ~~~~~~~~~~~~~~~~~
               811:  ret
    next:2862'0     ~~~~~
               812: .Lfunc_end76:
    next:2862'0     ~~~~~~~~~~~~~~
               813:  .size bswap_rotr_i64, .Lfunc_end76-bswap_rotr_i64
    next:2862'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
               814:  .cfi_endproc
    next:2862'0     ~~~~~~~~~~~~~~
               815:  # -- End function
    next:2862'0     ~~~~~~~~~~~~~~~~~~~
               816:  .globl bswap_rotl_i64 # -- Begin function bswap_rotl_i64
    next:2862'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
               817:  .p2align 2
    next:2862'0     ~~~~~~~~~~~~
               818:  .type bswap_rotl_i64,@function
    next:2862'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
               819: bswap_rotl_i64: # @bswap_rotl_i64
    next:2862'0     ~~~~~~~~~~~~~~~
               820:  .cfi_startproc
               821: # %bb.0:
    next:2905'0             X error: no match found
               822:  rev8 a0, a0
    next:2905'0     ~~~~~~~~~~~~~
    next:2905'1      ?            possible intended match
               823:  rori a0, a0, 32
    next:2905'0     ~~~~~~~~~~~~~~~~~
               824:  ret
    next:2905'0     ~~~~~
               825: .Lfunc_end77:
    next:2905'0     ~~~~~~~~~~~~~~
               826:  .size bswap_rotl_i64, .Lfunc_end77-bswap_rotl_i64
    next:2905'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
               827:  .cfi_endproc
    next:2905'0     ~~~~~~~~~~~~~~
                 .
                 .
                 .
    >>>>>>

    --

    ********************

    ********************
    Failed Tests (2):
      LLVM :: CodeGen/RISCV/rv32zbp.ll
      LLVM :: CodeGen/RISCV/rv64zbp.ll


    Testing Time: 1027.64s
      Skipped          :    20
      Unsupported      :  1596
      Passed           : 47432
      Expectedly Failed:   161
      Failed           :     2

    1 warning(s) in tests
    make[3]: *** [CMakeFiles/check-all.dir/build.make:71: CMakeFiles/check-all] Error 1
    make[3]: Leaving directory '/build/llvm-src-15.0.6/llvm/build'
    make[2]: *** [CMakeFiles/Makefile2:9246: CMakeFiles/check-all.dir/all] Error 2
    make[2]: Leaving directory '/build/llvm-src-15.0.6/llvm/build'
    make[1]: *** [CMakeFiles/Makefile2:9253: CMakeFiles/check-all.dir/rule] Error 2
    make[1]: Leaving directory '/build/llvm-src-15.0.6/llvm/build'
    make: *** [Makefile:228: check-all] Error 2

</details>